### PR TITLE
Feat/any collection

### PR DESCRIPTION
**Pull request by cbellot000**
- rename existing `Collection` base class as `CollectionBase`
- Create API for collection: registering new type of collection can be done in `src/ansys/dpf/core/__init__.py` using `_CollectionFactory`.
  Internal implementation uses AnyCollection
- Add API to create/cast any from `AnyCollection` and `CustomTypeField`

### CODE REVIEW

File: src/ansys/dpf/core/collection_base.py [hunk 1]
```diff
@@ -0,0 +1,679 @@
+"""
+Collection
```
Comment (cbellot000) [inline src/ansys/dpf/core/collection_base.py:2]:
```suggestion
CollectionBase
```

File: src/ansys/dpf/core/core.py [hunk 2]
```diff
@@ -257,10 +257,8 @@ def _description(dpf_entity_message, server=None):
     -------
        description : str
     """
-    try:
-        return BaseService(server, load_operators=False)._description(dpf_entity_message)
-    except:
-        return ""
+    return BaseService(server, load_operators=False)._description(dpf_entity_message)
```
Comment (cbellot000) [inline src/ansys/dpf/core/core.py:260]:
```suggestion
    try:
        return BaseService(server, load_operators=False)._description(dpf_entity_message)
    except:
        return ""
```

**Comment by PProfizi** [comment]
Hey @cbellot000, could you please provide a description of the changes?

File: src/ansys/dpf/core/__init__.py [hunk 3]
```diff
@@ -103,6 +103,14 @@
 
 from ansys.dpf.core.dpf_operator import available_operator_names
 
+
+from ansys.dpf.core.collection import CollectionFactory as _CollectionFactory
+
+
```
Comment (cbellot000) [inline src/ansys/dpf/core/__init__.py:109]:
```suggestion

# register classes for collection types:
```

File: src/ansys/dpf/core/collection.py [hunk 4]
```diff
@@ -1,671 +1,105 @@
...
-        if list is not None:
-            self._set_integral_entries(list)
-
-    def create_subtype(self, obj_by_copy):
-        return str(obj_by_copy)
 
-    def _set_integral_entries(self, input):
-        for s in input:
-            self._api.collection_add_string_entry(self, s)
+def CollectionFactory(subtype, BaseClass=Collection):
```
Comment (cbellot000) [inline src/ansys/dpf/core/collection.py:100]:
```suggestion
def CollectionFactory(subtype, BaseClass=Collection):
    """Creates classes deriving from Collection at runtime for a given subtype."""
```

File: src/ansys/dpf/core/collection_base.py [hunk 5]
```diff
@@ -0,0 +1,679 @@
+"""
+CollectionBase
+===========
```
Comment (cbellot000) [inline src/ansys/dpf/core/collection_base.py:3]:
```suggestion
==============
```

File: src/ansys/dpf/core/collection_base.py [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+TYPE = TypeVar('TYPE')
+
+
+class CollectionBase(Generic[TYPE]):
+    type: Optional[type[TYPE]]
+    """Represents a collection of entries ordered by labels and IDs.
+
+    Parameters
+    ----------
+    dpf_type :
```
Comment (cbellot000) [inline src/ansys/dpf/core/collection_base.py:38]:
```suggestion
```

File: src/ansys/dpf/core/collection_base.py [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+
```
Comment (cbellot000) [inline src/ansys/dpf/core/collection_base.py:39]:
```suggestion
```

File: src/ansys/dpf/core/collection_base.py [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+    dpf_vector,
+    dpf_array,
+)
+from typing import Optional, Generic, TypeVar
+
+TYPE = TypeVar('TYPE')
+
+
+class CollectionBase(Generic[TYPE]):
+    type: Optional[type[TYPE]]
```
Comment (cbellot000) [inline src/ansys/dpf/core/collection_base.py:33]:
```suggestion
    type: Optional[type[TYPE]] # type of the entries in the collection.
```

File: src/ansys/dpf/core/common.py [hunk 9]
```diff
@@ -339,9 +339,21 @@ def type_to_internal_object_keyword():
         workflow,
         streams_container,
         generic_data_container,
+        any,
+        collection,
     )
 
-    return {
+    class _smart_dict_types(dict):
+        def __getitem__(self, item):
```
Comment (cbellot000) [inline src/ansys/dpf/core/common.py:347]:
```suggestion
        def __getitem__(self, item):
        """If found returns the item of key == ìtem`, else returns item with key matching `issubclass(item, key)`."""
```

**Comment by cbellot000** [comment]
> Hey @cbellot000, could you please provide a description of the changes?

yes, it's done

**Comment by PProfizi** [comment]
> * rename existing `Collection` base class as `CollectionBase`
> * Create API for collection: registering new type of collection can be done in `src/ansys/dpf/core/__init__.py` using `_CollectionFactory`.
>   Internal implementation uses AnyCollection
> * Add API to create/cast any from `AnyCollection` and `CustomTypeField`

The rename is not going to be retro-compatible, right? But I guess no-one should be using it anyway. Should we make the `CollectionBase` class private?

**Pull request merged**